File: src/zros_cpp/publisher_manager.cc
//
// Created by pairs on 7/25/18.
//

#include <zros/publisher_manager.h>
#include <zros/singleton.h>
#include <sspdlog/sspdlog.h>
#include "impl/publisher_impl.h"

namespace zros {
    void PublisherManager::publish(const std::string &topic, const std::string &content) {
        impl_->publish(topic, content);
    }

    PublisherManager::PublisherManager() {
        impl_ = std::make_shared<PublishersImpl>();
        service_discovery_ = SingleTon<ServiceDiscovery>::Instance();
    }

    bool PublisherManager::registerPublisher(const std::shared_ptr<IPublisher> publisher) {
        // 1. register to memory first (should bind port first)
        impl_->registerPublisher(publisher);
        // 2. register to master
        bool ok = service_discovery_->addPublisher(publisher);
        if (!ok) {
            SSPD_LOG_WARNING << "register publisher " << publisher->get_topic() << " to master failed";
            return false;
        }
        return true;
    }

    bool PublisherManager::unregisterPublisher(const std::string &topic) {
        return false;
    }
}

File: src/zros_python/publisher_manager.py
# -*- coding: utf-8 -*-
""" The publisher management

data: 2018-08-01
author: pairs
"""
import zros_python.impl.publisher_impl as impl
import zros_python.zros as zros
import logging

logger = logging.getLogger(__name__)


class PublisherManager(object):
    """

    """
    def __init__(self):
        self._impl = impl.PublishersImpl()
        self._gsd = zros.get_service_discovery()

    def register_publisher(self, publisher):
        self._impl.register_publisher(publisher)
        ok = self._gsd.add_publisher(publisher)
        if ok is False:
            logger.warning(u'register publisher %s to master failed', publisher.get_topic())
            return False
        return True

    def unregister_publisher(self, topic):
        pass

    def publish(self, topic, content):
        self._impl.publish(topic, content)
